### PR TITLE
Array index with constant index: Avoid going via vanilla representation

### LINKED ISSUE #3895: Optimize array index with small indices
@timohanke said that one array access costs 71 instructions. I wonder how much of that is spent on converting the index to `Nat` (if it isn’t already a `Nat`) and then converting it back to an unboxed number. Maybe we can optimize some patterns.

Certainly for constant indices like in
https://github.com/research-ag/motoko-lib/blob/4f49165aa1a0fdbb18911184c12decd11d059cdb/src/Sha256.mo#L209-L216
there should be an easy win in code like
```
    compile_exp_vanilla env ae e1 ^^ (* offset to array *)
    compile_exp_vanilla env ae e2 ^^ (* idx *)
    Arr.idx_bigint env ^^
```
where we can use `compile_exp env ae e2` to get the stackrep, and then if it is constant use `Arr.idx` and pass that constant.

### PR DESCRIPTION
**Pull request by nomeata**
There is some benefit. Let’s see what the perf report will say
```diff
           local.get $s
-          i32.const 14
-          call $Array.idx_bigint
+          i32.const 7
+          call $Array.idx
           local.get $s
-          i32.const 14
-          call $Array.idx_bigint
+          i32.const 7
+          call $Array.idx
           i32.load offset=1
```
Fixes #3895.

### CODE REVIEW

File: src/codegen/compile.ml [hunk 1]
```diff
@@ -8812,7 +8812,24 @@ let rec compile_lexp (env : E.t) ae lexp =
...
+*)
+and compile_array_index env ae e1 e2 =
+  let code_arr = compile_exp_vanilla env ae e1 in (* offset to array *)
+  let sr, code_idx = compile_exp env ae e2 in (* index *)
+  match sr with
+  (* If we have a (small) constant index, do not convert to Vanilla and back *)
+  (* I am not bothering with numbers above 2^31-1, where Ocamls'
+     int32_of_big_int would do the wrong thing *)
+  | SR.Const (_, Const.Lit (Const.BigInt n))
+    when Big_int.le_big_int n (Big_int.big_int_of_int 0x7fffffff)->
```
Comment (ggreif) [inline src/codegen/compile.ml:8826]:
```suggestion
    when Big_int.(le_big_int n (big_int_of_int 0x7fffffff)) ->
```

File: src/codegen/compile.ml [hunk 2]
```diff
@@ -8812,7 +8812,24 @@ let rec compile_lexp (env : E.t) ae lexp =
...
-and compile_prim_invocation (env : E.t) ae p es at =
+(* Common code for a[e] as lexp and as exp.
+Traps or puts the pointer to the field on the stack
+*)
+and compile_array_index env ae e1 e2 =
+  let code_arr = compile_exp_vanilla env ae e1 in (* offset to array *)
+  let sr, code_idx = compile_exp env ae e2 in (* index *)
+  match sr with
+  (* If we have a (small) constant index, do not convert to Vanilla and back *)
+  (* I am not bothering with numbers above 2^31-1, where Ocamls'
```
Comment (ggreif) [inline src/codegen/compile.ml:8823]:
```suggestion
  (* I am not bothering with numbers above 2^31-1, where Ocaml's
```

**Review by ggreif — DISMISSED** [review]
This looks like a valid refactoring to me, should be okay to go in!

**Comment by nomeata** [comment]
The refactoring is in https://github.com/dfinity/motoko/pull/3897, though :-)

I'm away from the laptop for the weekend, so feel free to fix the nits there and merge.

**Comment by luc-blaeser** [comment]
Just doing some repo cleanup for better focussing on active work:
Closing PR as it seems inactive for a longer time. Please re-open if it is still relevant and after it has been updated with latest master changes.

**Pull request closed without merging**